src/GroupManager.php

```diff
@@ -93,6 +101,10 @@ public function addGroup($entity_type_id, $bundle_id) {
     $editable->set('groups', $groups);
     $saved = $editable->save();
 
+    // Notify other module we added a new group.
+    // todo: should this be an event?
+    $this->moduleHandler->invokeAll('og_group_created', [$entity_type_id, $bundle_id]);
```

Comment:
Answering @chx from above:

> 1. It is easy to call a class member from a hook.

Which is a workaround and you would have a real OO implementation (i.e. an event listener) after all which should be used by your OO code that you want to unit test. So the hook would be only a wrapper which will not even be used by your own code. Using an event listener is a structural solution.

> 2. Then you will test said class

Workaround.

> Same

Workaround.

> So it's not a problem as it has not been for 15 years

It is a problem, unless you want to claim you never went looking for non-existing hooks in module.api.php files. Also not everybody uses PHPStorm. With event listeners you will never have to look for the implementations because you can readily find them.

> That is incredibly rare

But it happens and event listeners are a structural solution to this terrible Drupalism.

> There is compelling reason to do so, have you seen the amount of boilerplate an event listener requires?

Yes, it requires very little. You only need to declare the event you are listening to. This is everything:

```
    /**
     * {@inheritdoc}
     */
    public static function getSubscribedEvents()
    {
        return [MyEvent::EVENT_NAME => 'myListener'];
    }
```

> That's because core strictly delineates between sane people using hooks and the crazy people drunk on the Symfony kool-aid.

This is not only a Symfony thing, this is the observer pattern which has been a staple in OO programming for a long time before Symfony even existed.